libp2p/network/health/monitor.py

```diff
@@ -0,0 +1,360 @@
+                if initial_delay and initial_delay > 0:
+                    nursery.start_soon(self._sleep_then_start, initial_delay)
+                else:
+                    nursery.start_soon(self._monitor_connections_task)
+                self._monitoring_task_started.set()
+
+                # Wait until cancelled
+                await trio.sleep_forever()
+
+        except trio.Cancelled:
+            logger.info("ConnectionHealthMonitor service cancelled")
+            self._stop_monitoring.set()
+            raise
+
+    async def _sleep_then_start(self, delay: float) -> None:
+        try:
+            await trio.sleep(delay)
+        finally:
+            # Start monitoring after delay; nursery cancellation handles shutdown
+            await self._monitor_connections_task()
+
+    async def _monitor_connections_task(self) -> None:
+        """Main monitoring loop that runs periodic health checks."""
+        logger.info(
+            f"Health monitoring started with "
+            f"{self.config.health_check_interval}s interval"
+        )
+
+        try:
+            while True:
+                # Wait for either the check interval or stop signal
+                with trio.move_on_after(self.config.health_check_interval):
+                    await self._stop_monitoring.wait()
+                    break  # Stop signal received
+
+                # Perform health checks on all connections
+                await self._check_all_connections()
+
+        except trio.Cancelled:
+            logger.info("Health monitoring task cancelled")
+            raise
+        except Exception as e:
+            logger.error(f"Health monitoring task error: {e}", exc_info=True)
+            raise
+
+    async def _check_all_connections(self) -> None:
+        """Check health of all connections across all peers."""
+        try:
+            # Get snapshot of current connections to avoid modification during iteration
+            current_connections = self.swarm.connections.copy()
+
+            for peer_id, connections in current_connections.items():
+                if not connections:
+                    continue
+
+                # Check each connection to this peer
+                for conn in list(connections):  # Copy list to avoid modification issues
+                    try:
+                        await self._check_connection_health(peer_id, conn)
+                    except Exception as e:
+                        logger.error(f"Error checking connection to {peer_id}: {e}")
+
+        except Exception as e:
+            logger.error(f"Error in connection health check cycle: {e}")
+
+    async def _check_connection_health(self, peer_id: ID, conn: INetConn) -> None:
+        """Check health of a specific connection."""
+        try:
+            # Skip checks during connection warmup window
+            warmup = getattr(self.config, "health_warmup_window", 0.0)
+            if warmup and hasattr(conn, "established_at"):
```

Comment:
This line checks if an `INetConn` has an `established_at` attribute, but that attribute is never defined for either `INetConn` or `SwarmConn`, rendering the code in this section effectively dead.